### PR TITLE
[jit] Fix weak module cuda() _flat_weights bug

### PR DESCRIPTION
**Pull request by driazati**
Dynamically creating a type at runtime was messing up the MRO and has been causing many other problems. I think it's best to delete it, this causes a regression since
```python
self.linear = nn.Linear(10, 10)
isinstance(self.linear, nn.Linear)
```
will now be `False` again, but this will be fixed once recursive script mode is the default (#20939)

Differential Revision: [D15560549](https://our.internmc.facebook.com/intern/diff/15560549/)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -6431,16 +6431,16 @@ def tensor_test(x, y):
...
-    
-    def test_number_all(self):	
-        def int1():	
-            return all(torch.tensor([1,2,3],dtype=torch.uint8))	
-        def int2():	
-            return all(torch.tensor([1,0,3],dtype=torch.uint8))	
-
-        self.checkScript(int1, ())	
+
+    def test_number_all(self):
```
Comment (eellison) [inline test/test_jit.py]: is this meant to be in here?

File: test/test_jit.py [hunk 2]
```diff
@@ -11922,6 +11922,8 @@ def forward(self, x):
         weak_mod.weight = torch.nn.Parameter(torch.ones(5, 5) * 100)
         self.assertFalse(strong_mod(inp).allclose(weak_mod(inp)))
 
+    @unittest.skipIf(hasattr(torch.jit, 'WeakScriptModuleProxy'), "# TODO: re-enable"
```
Comment (eellison) [inline test/test_jit.py]: re-enable this when what lol

**Review by eellison — COMMENTED** [review]
Do you mind explaining what MRO was and how exactly htis fixed the cuda() bug (and maybe add test for that bug) ?

File: test/test_jit.py (hunk 2 above)
Comment (driazati) [reply test/test_jit.py]: The full message is split over 2 strings

**Comment by driazati** [comment]
For one it was making an inheritance diamond on `nn.Module` and was letting you use non-ScriptModule stuff it inherited on something that should have been a `ScriptModule`, so the changes were all happening the Python side.

**Comment by eellison** [comment]
Yea the CUDA bug seems worse than the isinstance bug. Is this PR workable instead https://github.com/pytorch/pytorch/pull/19403 ? We should also re open https://github.com/pytorch/pytorch/pull/19403 if this lands

**Comment by eellison** [comment]
Could you add a test ?

**Review by eellison — APPROVED** [review]
LGTM with follow up to test. Once this lands we should re-open the isinstance bug

File: test/test_jit.py [hunk 3]
```diff
@@ -12688,6 +12690,16 @@ def fn(lst):
...
 
+    @unittest.skipIf(not RUN_CUDA, "no CUDA")
+    def test_weak_cuda(self):
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__()
+                self.lstm = torch.nn.LSTM(5, 5)
+                self.lstm.cuda()
+
+        m = M()
```
[the remark this answers is not in the document]
Comment (driazati) [reply test/test_jit.py]: What do you mean? This isn't testing the forward pass, just that a weak module can be made into a cuda module.

[the remark this answers is not in the document]
Comment (eellison) [reply test/test_jit.py]: Yes could you verify that the ouptut of the module is a cuda tensor. This is just testing that it doesn't error, it's not testing the runtime.

**Pull request closed without merging**